### PR TITLE
Allow schema_extra to be a classmethod

### LINKED ISSUE #1305: schema_extra should be allowed to be a classmethod
In my opinion, this check is unnecessarily restrictive:
https://github.com/samuelcolvin/pydantic/blob/938335c46f7e31da71e5c2a45eb3cf1ad51030b5/pydantic/schema.py#L470-L471
Also, it raises `TypeError` only for classmethods, but not for regular methods.

### PR DESCRIPTION
**Pull request by laevilgenius**
## Change Summary
Allow `schema_extra` to be a classmethod.

## Related issue number
Resolves #1305

## Checklist

* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI and coverage remains at 100%
* [ ] Documentation reflects the changes where applicable
* [ ] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

**Comment by laevilgenius** [comment]
I find it far easier to remove the check altogether, than tinker with method inspection on class object. Not sure what is the best way to make mypy happy though.
@samuelcolvin wdyt?

File: tests/test_schema.py [hunk 1]
```diff
@@ -1534,20 +1534,6 @@ def schema_extra(schema):
     assert Model.schema() == {'title': 'Model', 'type': 'override'}
 
 
-def test_model_with_schema_extra_callable_classmethod_asserts():
```
Comment (samuelcolvin) [inline tests/test_schema.py]: let's keep this test but demonstrate that it works.

Comment (laevilgenius) [reply tests/test_schema.py]: Done.

**Comment by laevilgenius** [comment]
So, I've added a test for instance methods, but I'm not sure how useful it is.

Lint is failing due to obsolete pyflakes version (fixed in master, but not yet released on pypi). How should we proceed?

File: pydantic/main.py [hunk 2]
```diff
@@ -52,6 +54,19 @@
     ConfigType = Type['BaseConfig']
     Model = TypeVar('Model', bound='BaseModel')
 
+    class SchemaExtraCallable(typing_extensions.Protocol):
+        @overload
+        def __call__(self, schema: Dict[str, Any]) -> None:
+            pass
+
+        @overload
+        def __call__(self, schema: Dict[str, Any], model: Type['Model']) -> None:
```
Comment (samuelcolvin) [inline pydantic/main.py]: Surely you don't need both this case and, the one below?

Better to use `model_class` than `model` to be clear.

You need to add `. # noqa F811` to get linting to pass, see other examples in the code base

File: pydantic/main.py [hunk 3]
```diff
@@ -89,7 +104,7 @@ class BaseConfig:
     getter_dict: Type[GetterDict] = GetterDict
     alias_generator: Optional[Callable[[str], str]] = None
     keep_untouched: Tuple[type, ...] = ()
-    schema_extra: Union[Dict[str, Any], Callable[[Dict[str, Any]], None]] = {}
+    schema_extra: Union[Dict[str, Any], 'SchemaExtraCallable'] = {}
```
Comment (samuelcolvin) [inline pydantic/main.py]: maybe you could test this works by adding a to `tests/mypy/modules/success.py`?

Comment (laevilgenius) [reply pydantic/main.py]: I could use some help here. I've been trying something like this to pass
```py
class ConfigSchemaExtraCallable(BaseConfig['MyModel']):
    @staticmethod
    def schema_extra(schema: Dict[str, Any], model_class: Type['MyModel']) -> None:
        pass
```
to no avail.
Even when I made `BaseConfig` generic, mypy still complained about signature incompatibility.

File: pydantic/main.py (hunk 2 above)
Comment (laevilgenius) [reply pydantic/main.py]: Done.

**Comment by samuelcolvin** [comment]
spent ages trying to improve on this, but nothing seems to work.

Let's go with this.

**Pull request merged**